cpp/arcticdb/util/std_ranges_utils.hpp

```diff
@@ -0,0 +1,165 @@
+#pragma once
+
+#include <ranges>
+#include <unordered_map>
+#include <vector>
+#include <algorithm>
+#include <functional>
+#include <optional>
+#include <stack>
+#include <memory>
+
+namespace arcticdb::util {
+
+// Standard library replacement for folly::gen::groupBy
+template<typename Container, typename KeyFunc>
+auto group_by(Container&& container, KeyFunc&& key_func) {
+    // Use decltype to get the correct value type from the range
+    using ValueType = std::remove_reference_t<decltype(*std::begin(container))>;
+    using KeyType = std::invoke_result_t<KeyFunc, ValueType>;
+    
+    std::unordered_map<KeyType, std::vector<ValueType>> result;
```

Comment:
nit: an option to specify the map type as a template template parameter would be cool.